### ISSUE TITLE
termux incorrectly limits cursor movement to scroll region

### ISSUE THREAD

**Issue by nicm** [termux/termux-app#1340]
Hi

termux 0.80 on Android.

DECSTBM should only limit scrolling not cursor movement. Compare this in xterm and in termux (the xxx should be above the aaa):
~~~~
printf '\033[4;7r\033[3;1Haaa\033[Axxx' 
~~~~
Cursor movement should not be affected by the scroll region but termux appears to be forcing the cursor inside the region. I suspect this should be 0 rather than mTopMargin: https://github.com/termux/termux-app/blob/e5a8c0eb4af5e7f4b86fef000948386cb073f7db/terminal-emulator/src/main/java/com/termux/terminal/TerminalEmulator.java#L1379
And similarly for CUD.

**Comment by nicm** [comment]
Here is termux: 
![Screenshot_20191109-080319](https://user-images.githubusercontent.com/304243/68525231-72715980-02c7-11ea-8371-a7f6af6aa9b7.png)

**Comment by nicm** [comment]
And here is xterm: 
![xxx](https://user-images.githubusercontent.com/304243/68525202-3211db80-02c7-11ea-8829-f6a370ba1de2.png)

**Comment by KaaSHH** [comment]
> Hi
> 
> termux 0.80 on Android.
> 
> DECSTBM should only limit scrolling not cursor movement. Compare this in xterm and in termux (the xxx should be above the aaa):
> 
> ```
> printf '\033[4;7r\033[3;1Haaa\033[Axxx' 
> ```
> 
> Cursor movement should not be affected by the scroll region but termux appears to be forcing the cursor inside the region. I suspect this should be 0 rather than mTopMargin:
> 
> https://github.com/termux/termux-app/blob/e5a8c0eb4af5e7f4b86fef000948386cb073f7db/terminal-emulator/src/main/java/com/termux/terminal/TerminalEmulator.java#L1379
> 
> 
> And similarly for CUD.

**Comment by fornwall** [comment]
Thanks a lot for a good explanation and small test case!

Should be fixed by cdccc2c43350acdbf612117f2901e121f7b52938, which is now released in version 0.82 of the app.

Do you know if it's the same with DECSLRM right and left margins and CUF&CUB?

**Comment by nicm** [comment]
No, I didn't realise this before but it seems like left and right margins work differently and /do/ limit at least CUF, CUB, LF. Try this for example in xterm:
~~~~
printf '\033[?69h\033[10;20s\033[1;10Ha\033[3Db\nc'
sleep 10
~~~~
